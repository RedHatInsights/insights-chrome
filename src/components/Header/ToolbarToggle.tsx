import React, { useState } from 'react';
import { useAtomValue } from 'jotai';
import { Dropdown, DropdownItem, DropdownList } from '@patternfly/react-core/dist/dynamic/components/Dropdown';
import { MenuToggle } from '@patternfly/react-core/dist/dynamic/components/MenuToggle';
import { PopoverPosition } from '@patternfly/react-core/dist/dynamic/components/Popover';

import ChromeLink from '../ChromeLink/ChromeLink';
import { isPreviewAtom } from '../../state/atoms/releaseAtom';

export type ToolbarToggleDropdownItem = {
  url?: string;
  appId?: string;
  target?: string;
  title: string;
  onClick?: (event: MouseEvent | React.MouseEvent<any, MouseEvent> | React.KeyboardEvent<Element>) => void;
  isHidden?: boolean;
  isDisabled?: boolean;
  rel?: string;
};

export type ToolbarToggleProps = {
  icon?: React.ElementType;
  dropdownItems: ToolbarToggleDropdownItem[];
  widgetType?: string | number;
  className?: string;
  id?: string | number;
  hasToggleIndicator?: null;
  ouiaId?: string;
  isHidden?: boolean;
  ariaLabel?: string;
};

const ToolbarToggle = (props: ToolbarToggleProps) => {
  const [isOpen, setIsOpen] = useState(false);
  const isPreview = useAtomValue(isPreviewAtom);

  const onSelect = () => {
    setIsOpen((prev) => !prev);
  };

  const onToggle = () => setIsOpen((prev) => !prev);

  const onClickInternal = (
    ev: MouseEvent | React.KeyboardEvent<Element> | React.MouseEvent<any, MouseEvent>,
    url?: string,
    onClick?: ToolbarToggleDropdownItem['onClick']
  ) => {
    ev.preventDefault();
    if (url) {
      window.location.href = `${url}`;
    }

    if (onClick) {
      onClick(ev);
    }
  };

  // Render the question mark icon items
  const dropdownItems = props.dropdownItems.map(
    ({ url, appId, title, onClick, isHidden, isDisabled, target = '_blank', rel = 'noopener noreferrer', ...rest }) =>
      !isHidden ? (
        <DropdownItem
          key={title}
          ouiaId={title}
          isDisabled={isDisabled}
          component={
            appId && url
              ? ({ className: itemClassName }) => (
                  <ChromeLink {...rest} className={itemClassName} href={url} target={target} rel={rel} isBeta={isPreview} appId={appId}>
                    {title}
                  </ChromeLink>
                )
              : url
              ? 'a'
              : 'button'
          }
          // Because the urls are using 'a', don't use onClick for accessibility
          // If it is a button, use the onClick prop
          {...(appId
            ? {}
            : url
            ? {
                href: url,
                target,
                rel,
                ...rest,
              }
            : { onClick: (ev) => onClickInternal(ev, url, onClick) })}
        >
          {title}
        </DropdownItem>
      ) : (
        <React.Fragment key="fragment" />
      )
  );

  return (
    <Dropdown
      popperProps={{
        position: PopoverPosition.right,
      }}
      onOpenChange={setIsOpen}
      toggle={(toggleRef) => (
        <MenuToggle
          ref={toggleRef}
          variant={props.icon ? 'plain' : 'default'}
          className={props.className}
          id={props.id?.toString()}
          onClick={onToggle}
          aria-label={props.ariaLabel}
          isExpanded={isOpen}
          icon={props.icon && <props.icon />}
        />
      )}
      isOpen={isOpen}
      onSelect={onSelect}
      ouiaId={props.ouiaId}
    >
      <DropdownList>{dropdownItems}</DropdownList>
    </Dropdown>
  );
};

export default ToolbarToggle;
