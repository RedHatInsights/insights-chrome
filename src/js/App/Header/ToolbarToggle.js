import React, { Component } from 'react';

import { DropdownToggle, DropdownItem, DropdownPosition, Dropdown } from '@patternfly/react-core';
import PropTypes from 'prop-types';
import ChromeLink from '../Sidenav/Navigation/ChromeLink';
import { isBeta } from '../../utils';

class ToolbarToggle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
    };
    this.onSelect = this.onSelect.bind(this);
    this.onToggle = this.onToggle.bind(this);
  }

  onSelect() {
    this.setState({ isOpen: !this.state.isOpen });
  }

  onToggle(isOpen) {
    this.setState({ isOpen });
  }

  onClick(ev, url, onClick) {
    ev.preventDefault();
    if (url) {
      window.location.href = `${url}`;
    }

    if (onClick) {
      onClick(ev);
    }
  }

  render() {
    // Render the questionmark icon items
    const dropdownItems = this.props.dropdownItems.map(
      ({ url, appId, title, onClick, isHidden, target = '_blank', rel = 'noopener noreferrer', ...rest }) =>
        !isHidden ? (
          <DropdownItem
            key={title}
            ouiaId={title}
            component={
              appId ? (
                <ChromeLink href={url} target={target} rel={rel} isBeta={isBeta()} appId={appId}>
                  {title}
                </ChromeLink>
              ) : url ? (
                'a'
              ) : (
                'button'
              )
            }
            // Because the urls are using 'a', don't use onClick for accessibility
            // If it is a button, use the onClick prop
            {...(appId
              ? {}
              : url
              ? {
                  href: url,
                  target,
                  rel,
                  ...rest,
                }
              : { onClick: (ev) => this.onClick(ev, url, onClick) })}
          >
            {title}
          </DropdownItem>
        ) : (
          <React.Fragment key="fragment" />
        )
    );

    const toggle = (
      <DropdownToggle
        className={this.props.className}
        id={this.props.id}
        ouiaId={this.props.id}
        toggleIndicator={this.props.hasToggleIndicator}
        onToggle={this.onToggle}
      >
        {this.props.icon && <this.props.icon />}
      </DropdownToggle>
    );

    return (
      <Dropdown
        position={DropdownPosition.right}
        toggle={toggle}
        isOpen={this.state.isOpen}
        dropdownItems={dropdownItems}
        onSelect={this.onSelect}
        ouiaId={this.props.ouiaId}
        isPlain
      />
    );
  }
}

ToolbarToggle.propTypes = {
  icon: PropTypes.elementType,
  dropdownItems: PropTypes.array,
  widgetType: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  className: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hasToggleIndicator: PropTypes.bool,
  ouiaId: PropTypes.string,
  isHidden: PropTypes.bool,
};

export default ToolbarToggle;
