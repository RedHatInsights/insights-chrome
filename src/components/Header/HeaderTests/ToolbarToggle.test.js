import React from 'react';
import { act, fireEvent, render, screen } from '@testing-library/react';
import ToolbarToggle from '../ToolbarToggle';

describe('ToolbarToggle', () => {
  const clickSpy = jest.fn();
  const dropdownItems = [
    {
      url: 'url1',
      title: 'title1',
      onClick: clickSpy,
    },
    {
      title: 'title2',
      onClick: clickSpy,
    },
    {
      isHidden: true,
      title: 'title3',
      onClick: clickSpy,
    },
  ];
  const toolbarToggleProps = {
    dropdownItems,
    id: 'foo',
  };

  afterEach(() => {
    clickSpy.mockReset();
  });

  it('should render correctly', async () => {
    const { container } = render(<ToolbarToggle {...toolbarToggleProps} />);
    const toggleButton = container.querySelector('#foo');
    expect(toggleButton).toBeTruthy();
    await act(async () => {
      await fireEvent.click(toggleButton);
    });
    expect(container.querySelector('div')).toMatchSnapshot();
  });

  it('should open/close menu correctly', async () => {
    const { container } = render(<ToolbarToggle {...toolbarToggleProps} />);
    const toggleButton = container.querySelector('#foo');
    const expectedTexts = toolbarToggleProps.dropdownItems.filter((item) => !item.isHidden);
    expect(toggleButton).toBeInTheDocument();
    await act(async () => {
      await fireEvent.click(toggleButton);
    });

    // wait for async actions on toggle to complete
    await act(async () => {
      await Promise.resolve();
    });

    for (const item of expectedTexts) {
      expect(screen.getByText(item.title)).toBeInTheDocument();
    }
    // closes button
    await act(async () => {
      await fireEvent.click(toggleButton);
    });

    // wait for async actions on toggle to complete
    await act(async () => {
      await Promise.resolve();
    });
    for (const item of expectedTexts) {
      expect(screen.queryByText(item.title)).not.toBeInTheDocument();
    }
    // expect(container.querySelectorAll('.pf-v6-c-menu__list-item')).toHaveLength(0);
  });

  it('should call onClick menu item callback', async () => {
    const { container } = render(<ToolbarToggle {...toolbarToggleProps} />);
    const toggleButton = container.querySelector('#foo');
    await act(async () => {
      await fireEvent.click(toggleButton);
    });
    const actionButton = container.querySelector('button.pf-v6-c-menu__item');
    expect(actionButton).toBeTruthy();
    await act(async () => {
      await fireEvent.click(actionButton);
    });
    expect(clickSpy).toHaveBeenCalled();
  });
});
