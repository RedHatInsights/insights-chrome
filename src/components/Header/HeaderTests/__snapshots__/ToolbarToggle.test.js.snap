// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarToggle should render correctly 1`] = `
<div
  class="pf-v5-c-menu"
  data-ouia-component-id="OUIA-Generated-Dropdown-1"
  data-ouia-component-type="PF5/Dropdown"
  data-ouia-safe="true"
  data-popper-escaped="true"
  data-popper-placement="bottom-end"
  data-popper-reference-hidden="true"
  style="position: absolute; left: 0px; top: 0px; z-index: 9999; opacity: 1; transition: opacity 0ms cubic-bezier(.54, 1.5, .38, 1.11); min-width: 0px; transform: translate(0px, 0px);"
>
  <div
    class="pf-v5-c-menu__content"
  >
    <ul
      class="pf-v5-c-menu__list"
      role="menu"
    >
      <li
        class="pf-v5-c-menu__list-item"
        data-ouia-component-id="title1"
        data-ouia-component-type="PF5/DropdownItem"
        data-ouia-safe="true"
        href="url1"
        role="none"
      >
        <a
          class="pf-v5-c-menu__item"
          rel="noopener noreferrer"
          role="menuitem"
          tabindex="0"
          target="_blank"
        >
          <span
            class="pf-v5-c-menu__item-main"
          >
            <span
              class="pf-v5-c-menu__item-text"
            >
              title1
            </span>
          </span>
        </a>
      </li>
      <li
        class="pf-v5-c-menu__list-item"
        data-ouia-component-id="title2"
        data-ouia-component-type="PF5/DropdownItem"
        data-ouia-safe="true"
        role="none"
      >
        <button
          class="pf-v5-c-menu__item"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <span
            class="pf-v5-c-menu__item-main"
          >
            <span
              class="pf-v5-c-menu__item-text"
            >
              title2
            </span>
          </span>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
